[Docs] Distinguish warm vs cold vs frozen tiers

Leaf-Lin opened the pull request:
Today, the description at the top of this data tier page can be confusing for users on how to choose between warm vs cold or cold vs frozen. 

I added some descriptions around cold and frozen tiers to emphasise the ability of use of searchable snapshot and their storage benefit based on https://www.elastic.co/elasticsearch/elasticsearch-searchable-snapshots and https://www.elastic.co/blog/introducing-elasticsearch-frozen-tier-searchbox-on-s3. Feel free to reword it.

Leaf-Lin committed:
distinguish warm/cold vs frozen

Adding additional description around cold and frozen tiers to emphasise the ability of use of searchable snapshot and their storage benefit.

dakrone commented on docs/reference/datatiers.asciidoc:
```diff
@@ -10,8 +10,8 @@ typically share the same hardware profile:
 and rarely needs to be updated.
-* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated.
-* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated, kept in searchable snapshots.
+* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> are kept on cold tier eliminating the need for replicas with approximately 50% storage reduction compares to warm tiers.
```
I think this implies that it happens automatically, so I would recommend something like:
```suggestion
* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> may be kept on the cold tier, eliminating the need for replicas with approximately 50% storage reduction compared to the warm tier.
```

dakrone commented on docs/reference/datatiers.asciidoc:
```diff
@@ -10,8 +10,8 @@ typically share the same hardware profile:
-* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated.
-* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated, kept in searchable snapshots.
+* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> are kept on cold tier eliminating the need for replicas with approximately 50% storage reduction compares to warm tiers.
+* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated. <<partially-mounted,Partially mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> are kept on frozen tier extend the storage capacity even further by 20 times compares to warm tiers.
```
```suggestion
* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated. <<partially-mounted,Partially mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> are kept on frozen tier to extend the storage capacity even further by up to 20 times compare to the warm tier.
```

Leaf-Lin replied:
One thing I never grasp with English grammar is `tense`.  

In the previous [commit](https://github.com/elastic/elasticsearch/commit/5c97e845db636d3df4dd116c5519b4a2c0512e73#r809549540), you propose to change `compares` to `compared`.  In this sentence, you propose to change `compares` to `compare`.  

What have I missed? :P

dakrone replied:
I had a typo, they should both be `compared`, I fixed my typo now :)

dakrone reviewed (commented):
I left some wording recommendations but it would be good for @jrodewig to take a look also

Leaf-Lin committed:
Update docs/reference/datatiers.asciidoc

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

Leaf-Lin committed:
Update docs/reference/datatiers.asciidoc

One thing I never grasp with English grammar is `tense`.  

In the previous [commit](https://github.com/elastic/elasticsearch/pull/84155/commits/5c97e845db636d3df4dd116c5519b4a2c0512e73#r809549540), you propose to change `compares` to `compared`.  In this sentence, you propose to change `compares` to `compare`.  

What have I missed? :P

Co-authored-by: Lee Hinman <dakrone@users.noreply.github.com>

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

jrodewig commented on docs/reference/datatiers.asciidoc:
```diff
@@ -10,8 +10,8 @@ typically share the same hardware profile:
 and rarely needs to be updated.
-* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated.
-* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated, kept in searchable snapshots.
+* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> may be kept on the cold tier, eliminating the need for replicas with approximately 50% storage reduction compared to the warm tier.
```
Two suggestions:

- I'd point out that the storage savings is in disk space.
- I'd point out that you can use regular indices on the cold tier if wanted.

```suggestion
* <<cold-tier,Cold tier>> nodes hold time series data that is accessed
infrequently and not normally updated. To save space, you can keep
<<fully-mounted,fully mounted indices>> of
<<ilm-searchable-snapshot,{search-snaps}>> on the cold tier. These fully mounted
indices eliminate the need for replicas, reducing required disk space by
approximately 50% compared to the regular indices.
+
Alternatively, you can the cold tier to store regular indices with replicas instead
of using {search-snaps}. This lets you store older data on less expensive hardware
but doesn't reduce required disk space compared to the warm tier.
```

Leaf-Lin replied:
🤦 , I somehow missed this suggestion while I was working my way through the above.  I have added this commit in https://github.com/elastic/elasticsearch/pull/84155#discussion_r813521447 and in the main body.
 
To avoid duplication, I'd prefer to simplify the cold tier description in the summary at the top, and put the detailed description in the content, but I felt all of these points are important though, um...

Additionally, I am reading https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html and felt that some of the descriptions around cold or frozen data **nodes** can be updated accordingly or linked to content on this page.  Can you assist with that please?

jrodewig reviewed (commented):
Thanks @Leaf-Lin. I left a few comments and suggestions I think we should address before merging.

Let me know if you have any questions or if I can help in any way.

Leaf-Lin committed:
Update cold tier description

Leaf-Lin committed:
Update datatiers.asciidoc

Clarify frozen node needs a snapshot repo.

Leaf-Lin committed:
Update datatiers.asciidoc

Fix indentation

Leaf-Lin commented on docs/reference/datatiers.asciidoc:
```diff
@@ -75,13 +84,12 @@ For resiliency, indices in the warm tier should be configured to use one or more
-for replicas. The cold tier is still a responsive query tier, but data becomes
-read-only.
+Cold tier can be used for storing rarely updated time series data similar to 
+the warm tier.  It can also be used to for storing read-only
```
```suggestion
the warm tier.  It can also be used for storing read-only
```

Leaf-Lin committed:
Update docs/reference/datatiers.asciidoc

Leaf-Lin commented on docs/reference/datatiers.asciidoc:
```diff
@@ -10,8 +10,8 @@ typically share the same hardware profile:
 and rarely needs to be updated.
-* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated.
-* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated, kept in searchable snapshots.
+* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> may be kept on the cold tier, eliminating the need for replicas with approximately 50% storage reduction compared to the warm tier.
```
Great suggestion. Could you check if my latest commit satisfies the above?

I made quite a lot of modifications in the cold tier paragraph. Mainly to point out that one can use cold tier just like warm tier, or use searchable snapshot. 

```
Cold tier can be used for storing rarely updated time series data similar to 
the warm tier.  It can also be used for storing read-only
<<fully-mounted,fully mounted indices>> of
<<ilm-searchable-snapshot,{search-snaps}>>.  When one chooses to use searchable
snapshot on the cold tier, it eliminates the need for replicas with
approximately 50% storage reduction compared to the warm tier.
```

This is indeed a frequent question users are confused about the cold tier.

Leaf-Lin commented on docs/reference/datatiers.asciidoc:
```diff
@@ -75,13 +84,12 @@ For resiliency, indices in the warm tier should be configured to use one or more
-<<ilm-searchable-snapshot,{search-snaps}>> for resiliency, eliminating the need
-for replicas. The cold tier is still a responsive query tier, but data becomes
-read-only.
+Cold tier can be used for storing rarely updated time series data similar to 
```
```suggestion
Cold tier can be used for storing infrequently updated time series data similar to 
```
Change to infrequently to be consistent with other places.

Leaf-Lin committed:
Update docs/reference/datatiers.asciidoc

Leaf-Lin commented on docs/reference/datatiers.asciidoc:
```diff
@@ -10,8 +10,8 @@ typically share the same hardware profile:
-* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated.
-* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated, kept in searchable snapshots.
+* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated. <<fully-mounted,Fully mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> may be kept on the cold tier, eliminating the need for replicas with approximately 50% storage reduction compared to the warm tier.
+* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated. <<partially-mounted,Partially mounted indices>> of <<ilm-searchable-snapshot,{search-snaps}>> are kept on frozen tier to extend the storage capacity even further by up to 20 times compare to the warm tier.
```
I have modified the paragraph and used the word `exclusively`, hope that can justify explicitly mention?

```
* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed 
rarely and never updated. It is necessary to have a snapshot repository setup
in order to use frozen tier. The frozen tier stores
<<partially-mounted,Partially mounted indices>> of
<<ilm-searchable-snapshot,{search-snaps}>> exclusively.  This extends the
storage capacity even further by up to 20 times compare to the warm tier. 
```

Additionally, I have also seen quite a lot of confusion around frozen node setup where users may not realize it depends on a snapshot repository. So I added the sentence to emphasize the need.

Leaf-Lin commented on docs/reference/datatiers.asciidoc:
```diff
@@ -10,8 +10,17 @@ typically share the same hardware profile:
 and rarely needs to be updated.
-* <<cold-tier, Cold tier>> nodes hold time series data that is accessed infrequently and not normally updated.
-* <<frozen-tier, Frozen tier>> nodes hold time series data that is accessed rarely and never updated, kept in searchable snapshots.
+* <<cold-tier, Cold tier>> nodes hold time series data that is accessed 
+infrequently and not normally updated. <<fully-mounted,Fully mounted
+indices>> of <<ilm-searchable-snapshot,{search-snaps}>> may be kept on the
+cold tier, eliminating the need for replicas with approximately 50% storage
+reduction compared to the warm tier.
```
```suggestion
* <<cold-tier,Cold tier>> nodes hold time series data that is accessed
infrequently and not normally updated. To save space, you can keep
<<fully-mounted,fully mounted indices>> of
<<ilm-searchable-snapshot,{search-snaps}>> on the cold tier. These fully mounted
indices eliminate the need for replicas, reducing required disk space by
approximately 50% compared to the regular indices.
+
Alternatively, you can use the cold tier to store regular indices with replicas instead
of using {search-snaps}. This lets you store older data on less expensive hardware
but doesn't reduce required disk space compared to the warm tier.
```

Leaf-Lin committed:
Update docs/reference/datatiers.asciidoc

Leaf-Lin committed:
Update datatiers.asciidoc

Leaf-Lin committed:
Update datatiers.asciidoc

jrodewig commented on docs/reference/datatiers.asciidoc:
```diff
@@ -90,6 +107,7 @@ read-only.
 Once data is no longer being queried, or being queried rarely, it may move from
 the cold tier to the frozen tier where it stays for the rest of its life.
 
+It is necessary to have a snapshot repository setup in order to use frozen tier.
```
Nit: rewording
```suggestion
The frozen tier requires a snapshot repository.
```

jrodewig reviewed (commented):
Thanks @Leaf-Lin. I left some suggestions — mostly to deduplicate the content. If my suggestions look okay to you, I think this is good to go.

> Additionally, I am reading https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-node.html and felt that some of the descriptions around cold or frozen data nodes can be updated accordingly or linked to content on this page. Can you assist with that please?

Sure thing. I'll handle that in a separate PR.

Edit: I raised a draft PR here: https://github.com/elastic/elasticsearch/pull/84346. That gives you an idea of where I'm heading, but I think we can do better at content reuse. That'll be easier to plumb once these changes are merged in.

Leaf-Lin committed:
Update docs/reference/datatiers.asciidoc

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

Leaf-Lin committed:
Update docs/reference/datatiers.asciidoc

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

Leaf-Lin committed:
Update docs/reference/datatiers.asciidoc

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

Leaf-Lin committed:
Update docs/reference/datatiers.asciidoc

Co-authored-by: James Rodewig <james.rodewig@elastic.co>

Leaf-Lin commented:
@jrodewig .  LGTM, thanks for your suggestions! I think this still needs approval before merging?

jrodewig reviewed (approved):
LGTM. I'll merge this in.

jrodewig commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/84435">](https://github.com/elastic/elasticsearch/pull/84435)|
|✅|8.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/84436">](https://github.com/elastic/elasticsearch/pull/84436)|
|✅|7.17|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/84437">](https://github.com/elastic/elasticsearch/pull/84437)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
